### PR TITLE
Add Sparse support for torch.eye()

### LINKED ISSUE #46705: torch.eye not supported for sparse layout
1. Call torch.eye with layout = torch.sparse_coo

Below is a code sample:

import torch

x = torch.eye(4, layout=torch.sparse_coo, device=torch.device('cuda'))
print(x)

## Actual behavior
Traceback (most recent call last):
  File "sparse_tensor.py", line 46, in <module>
    x = torch.eye(4, layout=torch.sparse_coo, device=torch.device('cuda'))
RuntimeError: Could not run 'aten::eye.m_out' with arguments from the 'SparseCUDATensorId' backend. 'aten::eye.m_out' is only available for these backends: [CPUTensorId, CUDATensorId, VariableTensorId].

## Expected behavior
The Sparse Tensor should get created and printed properly as below:

tensor(indices=tensor([[0, 1, 2, 3],
                       [0, 1, 2, 3]]),
       values=tensor([1., 1., 1., 1.]),
       size=(4, 4), nnz=4, layout=torch.sparse_coo)

## Environment

Versions of relevant libraries:
[pip3] numpy==1.19.1
[pip3] torch==1.8.0a0
[pip3] torchtext==0.8.0a0+7e267d2
[pip3] torchvision==0.8.0a0+be8192e
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               10.2.89              hfd86e86_1  
[conda] magma-cuda102             2.5.2                         1    pytorch
[conda] mkl                       2020.1                      217  
[conda] mkl-include               2020.1                      217  
[conda] mkl-service               2.3.0            py38he904b0f_0  
[conda] mkl_fft                   1.1.0            py38h23d657b_0  
[conda] mkl_random                1.1.1            py38h0573a6f_0  
[conda] numpy                     1.19.1           py38hbc911f0_0  
[conda] numpy-base                1.19.1           py38hfa32c7d_0  
[conda] torch                     1.8.0a0                   dev_0    <develop>
[conda] torchtext                 0.8.0a0+7e267d2          pypi_0    pypi
[conda] torchvision               0.8.0a0+be8192e          pypi_0    pypi

 - PyTorch Version: 1.8.0a0
 - OS: Linux
 - How you installed PyTorch (`conda`, `pip`, source): source
 - Build command you used (if compiling from source): python setup.py develop
 - Python version: 3.8
 - CUDA/cuDNN version: 10.2

### PR DESCRIPTION
**Pull request by hlahkar**
Fixes #46705

### CODE REVIEW

File: aten/src/ATen/native/sparse/SparseTensor.cpp [hunk 1]
```diff
@@ -125,6 +125,36 @@ Tensor sparse_coo_tensor(ArrayRef<int64_t> size, const TensorOptions& options) {
...
+  TORCH_CHECK(n >= 0, "n must be greater or equal to 0, got ", n);
+
+  if(m < 0) {
+    m = n;
+  }
+
+  // take the smaller of the rows/columns
+  int64_t sz = (m < n) ? m : n;
+
+  Tensor values = at::ones({sz});
```
Comment (ngimel) [inline aten/src/ATen/native/sparse/SparseTensor.cpp:143]: here you are creating values of potentially wrong dtype on a potentially wrong device, only to copy later. Create them with the right type.

File: aten/src/ATen/native/sparse/SparseTensor.cpp [hunk 2]
```diff
@@ -125,6 +125,36 @@ Tensor sparse_coo_tensor(ArrayRef<int64_t> size, const TensorOptions& options) {
...
+  // hard code sparse dimensions to 2 as this is a guranteed matrix
+  Tensor indices = at::empty({2, sz});
```
Comment (ngimel) [inline aten/src/ATen/native/sparse/SparseTensor.cpp:145]: You can do something like `indices = torch.arange(sz, options).unsqueeze(0).expand(2,-1)`, accessing by element as you do below is very inefficient.

**Review by ngimel — CHANGES_REQUESTED** [review]
Please add tests for this functionality.

File: aten/src/ATen/native/sparse/SparseTensor.cpp (hunk 1 above)
Comment (hlahkar) [reply aten/src/ATen/native/sparse/SparseTensor.cpp]: Have modified the same; using the dtype and device from the result Tensor

File: aten/src/ATen/native/sparse/SparseTensor.cpp (hunk 2 above)
Comment (hlahkar) [reply aten/src/ATen/native/sparse/SparseTensor.cpp]: Have incorporated the suggestions

**Comment by hlahkar** [comment]
> Please add tests for this functionality.

Added the tests for the same

File: test/test_sparse.py [hunk 3]
```diff
@@ -2177,6 +2177,36 @@ def test_factory_empty_indices(self):
         expected_indices = torch.empty((4, 0), dtype=torch.long, device=device)
         self.assertEqual(tensor._indices(), expected_indices)
 
+    def test_sparse_eye(self):
+        device = 'cuda' if self.is_cuda else 'cpu'
+
+        tensor = torch.eye(3, 4, layout=torch.sparse_coo, device=device)
```
Comment (ngimel) [inline test/test_sparse.py:2183]: Can you please call these tests in a loop for different datatypes? If you use (3,4), (4,3), (3,3) sizes your expects can remain the same throughout. 
Also, please test the `out` variant of the function (`torch.eye(3,3, layout=..., out=...)`). 
Also, please check the size of the result.

**Review by ngimel — COMMENTED** [review]
This looks much better, thank you! 
I have comments about the test, also, please edit documentation entry for `eye` to mention that it supports sparse layout.

File: test/test_sparse.py (hunk 3 above)
Comment (hlahkar) [reply test/test_sparse.py:2183]: Will update the test case as per the suggestion.

**Comment by hlahkar** [comment]
> This looks much better, thank you!
> I have comments about the test, also, please edit documentation entry for `eye` to mention that it supports sparse layout.

Can you please help in pointing out the correct documentation file; where I can update the same

File: test/test_sparse.py [hunk 4]
```diff
@@ -2177,6 +2177,44 @@ def test_factory_empty_indices(self):
...
+    def test_sparse_eye(self):
+        device = 'cuda' if self.is_cuda else 'cpu'
+        data_type = [torch.long, torch.int, torch.float, torch.half, torch.double, torch.short, torch.float64]
+
+        for i in range(7):
+            tensor = torch.eye(i + 2, i + 2, layout=torch.sparse_coo, device=device, dtype=data_type[i])
+            expected_indices = torch.arange(i + 2, device=device, dtype=torch.long).unsqueeze(0).expand(2, -1)
+            self.assertEqual(tensor._indices(), expected_indices)
+            expected_values = torch.ones(i + 2, device=device, dtype=data_type[i])
+            self.assertEqual(tensor._values(), expected_values)
```
Comment (ngimel) [inline test/test_sparse.py:2189]: This is a good start, you can roll all your tests from below into this loop
```suggestion
        for data_type in (torch.long, torch.int, torch.float, torch.half, torch.double, torch.short, torch.float64):
            size = 4
            expected_indices = torch.arange(size, device=device, dtype=torch.long).unsqueeze(0).expand(2, -1)
            expected_values = torch.ones(size, device=device, dtype=data_type)
            tensor = torch.eye(size, size, layout=torch.sparse_coo, device=device, dtype=data_type)
            # unfortunately, self.assertEqual does not do the right thing for sparse tensors, so mimic it
            self.assertEqual((tensor._indices(), tensor._values(), tensor.size()), (expected_indices, expected_values, torch.Size((size, size)))
            tensor = torch.eye(size, size+1, layout=torch.sparse_coo, device=device, dtype=data_type)
            self.assertEqual((tensor._indices(), tensor._values(), tensor.size()), (expected_indices, expected_values, torch.Size((size, size+1)))
            tensor = torch.eye(size-1, size, layout=torch.sparse_coo, device=device, dtype=data_type)
            self.assertEqual((tensor._indices(), tensor._values(), tensor.size()), (expected_indices, expected_values, torch.Size((size-1, size)))
            tensor = torch.eye(size, layout=torch.sparse_coo, device=device, dtype=data_type)
            self.assertEqual((tensor._indices(), tensor._values(), tensor.size()), (expected_indices, expected_values, torch.Size((size, size)))
            # + test out variants and that return is coalesced

            
            
```

File: aten/src/ATen/native/sparse/SparseTensor.cpp [hunk 5]
```diff
@@ -125,6 +125,35 @@ Tensor sparse_coo_tensor(ArrayRef<int64_t> size, const TensorOptions& options) {
   return at::_sparse_coo_tensor_with_dims(size.size(), 0, size, options.layout(at::kSparse));
 }
 
+/*eye init*/
+Tensor& eye_sparse(Tensor& result, int64_t n) {
+  return eye_sparse(result, n, -1);
```
Comment (ngimel) [inline aten/src/ATen/native/sparse/SparseTensor.cpp:130]: we no longer allow negative arguments, use `(result, n, n)` instead

File: aten/src/ATen/native/sparse/SparseTensor.cpp [hunk 6]
```diff
@@ -125,6 +125,35 @@ Tensor sparse_coo_tensor(ArrayRef<int64_t> size, const TensorOptions& options) {
...
+}
+
+Tensor& eye_sparse(Tensor& result, int64_t n, int64_t m) {
+  TORCH_CHECK(n >= 0, "n must be greater or equal to 0, got ", n);
+
+  if (m < 0) {
```
Comment (ngimel) [inline aten/src/ATen/native/sparse/SparseTensor.cpp:136]: negative args are not allowed, check that m is not negative

File: aten/src/ATen/native/sparse/SparseTensor.cpp [hunk 7]
```diff
@@ -125,6 +125,35 @@ Tensor sparse_coo_tensor(ArrayRef<int64_t> size, const TensorOptions& options) {
...
+  // take the smaller of the rows/columns
+  int64_t sz = (m < n) ? m : n;
+  TensorOptions options = TensorOptions(result.device());
+  options = options.dtype(result.dtype());
+
+  Tensor values = at::ones({sz}, options);
+
+  // hard code sparse dimensions to 2 as this is a guranteed matrix
+  options = options.dtype(ScalarType::Long);
+  Tensor indices = at::arange(sz, options).unsqueeze(0).expand({2, -1});
```
Comment (ngimel) [inline aten/src/ATen/native/sparse/SparseTensor.cpp:149]: call `.contiguous()` on indices, it's customary to return contiguous indices

File: aten/src/ATen/native/sparse/SparseTensor.cpp [hunk 8]
```diff
@@ -125,6 +125,35 @@ Tensor sparse_coo_tensor(ArrayRef<int64_t> size, const TensorOptions& options) {
...
+
+  get_sparse_impl(result)->resize_(2, 0, {n, m});
+  copy_into_sparse(result, indices, values, false);
```
Comment (ngimel) [inline aten/src/ATen/native/sparse/SparseTensor.cpp:152]: please set `coalesced` attribute to `true`

File: docs/source/sparse.rst [hunk 9]
```diff
@@ -445,6 +445,7 @@ tensors <sparse-coo-docs>`:
 :meth:`~torch.Tensor.dim`
 :meth:`~torch.Tensor.div`
 :meth:`~torch.Tensor.div_`
+:meth:`~torch.Tensor.eye`
```
Comment (ngimel) [inline docs/source/sparse.rst:448]: thanks for adding to the docs!

File: test/test_sparse.py (hunk 3 above)
Comment (ngimel) [reply test/test_sparse.py:2202]: Oops, sorry, I committed my suggestion accidentally, but  I did not actually check that it runs, feel free to revert

File: test/test_sparse.py [hunk 10]
```diff
@@ -2177,6 +2177,55 @@ def test_factory_empty_indices(self):
         expected_indices = torch.empty((4, 0), dtype=torch.long, device=device)
         self.assertEqual(tensor._indices(), expected_indices)
 
+    def test_sparse_eye(self):
+        device = 'cuda' if self.is_cuda else 'cpu'
+        data_type = [torch.long, torch.int, torch.float, torch.half, torch.double, torch.short, torch.float64]
+
+        for data_type in (torch.long, torch.int, torch.float, torch.half, torch.double, torch.short, torch.float64):
+            size = 4
+            expected_indices = torch.arange(size, device=device, dtype=torch.long).unsqueeze(0).expand(2, -1)
```
Comment (mruberry) [inline test/test_sparse.py:2186]: In addition to putting the tests below into this loop, compare torch.eye with a sparse layout with torch.eye with a strided layout by calling to_dense on the sparse result. Also assert that the sparse tensor returned from eye is coalesced with `is_coalesced`.

The current assertEquals that compare indices and values separately can be removed.

Comment (hlahkar) [reply test/test_sparse.py]: @mruberry - Updated the test case as per your suggestion, kindly review the same

**Pull request closed without merging**